src/components/Viewport.ts

```diff
@@ -1476,4 +1675,161 @@ export default class Viewport {
       } as Partial<NeedPanelEvent>,
     );
   }
+
+  private updateVisiblePanels() {
+    const state = this.state;
+    const options = this.options;
+    const isHorizontal = this.options.horizontal;
+    const cameraElement = this.cameraElement;
+    const visibleIndex = state.visibleIndex;
+
+    if (!options.renderOnlyVisible) {
+      return;
+    }
+
+    if (!this.nearestPanel) {
+      this.resetVisibleIndex();
+      while (cameraElement.firstChild) {
+        cameraElement.removeChild(cameraElement.firstChild);
+      }
+      return;
+    }
+
+    const panelManager = this.panelManager;
+    const allPanelCount = panelManager.getRange().max + 1;
+    const cloneCount = panelManager.getCloneCount();
+    const cameraPos = this.getCameraPosition();
+    const basePanel = this.nearestPanel;
+    const viewportBbox = this.getBbox();
+    const bbox = this.options.overflow
+      ? this.getWindowBBox()
+      : viewportBbox;
+    const viewportPos = isHorizontal
+      ? viewportBbox.x
+      : viewportBbox.y;
+
+    const checkLastPanel = (
```

Comment:
Do you have any reason to include checkLastPanel in updateVisiblePanels?

Reply:
Maybe It seems like parameter function has reference to above values.

Reply:
But it looks better to split this function into 2 or more part because it is hard to read.

Reply:
I think that was because that method is used only in this method.
Can I ask why that can be bad? Like, transpiled size?

Reply:
I've left the above comment before new comments were made.
I think that's nice, but I'm afraid it's a quite short function to split in half.
Can you show me how to do it?

Reply:
My thought is that `updateVisiblePanels()` is too long to read easy. and `checkLastPanel` is a part of issue(?). So I think it would be better to split functionality which have related.

>  it's not working code. It is just written in the flow of consciousness. And This is just code to convey my rough thinking.

```ts
private updateVisiblePanels() {
	const state = this.state;
	const options = this.options;
	const cameraElement = this.cameraElement;
	const visibleIndex = state.visibleIndex;

	if (!options.renderOnlyVisible) {
		return;
	}
	if (!this.nearestPanel) {
		this.resetVisibleIndex();
		while (cameraElement.firstChild) {
			cameraElement.removeChild(cameraElement.firstChild);
		}
		return;
	}

	const {lastPanelOfPrevDir, lastPanelOfNextDir} = this.getLastPanels();

	const newVisibleIndex = this.getNewVisibleIndex(lastPanelOfPrevDir, lastPanelOfNextDir);

	if (newVisibleIndex.min !== visibleIndex.min || newVisibleIndex.max !== visibleIndex.max) {
		state.visibleIndex = newVisibleIndex;
		if (isNaN(newVisibleIndex.min) || isNaN(newVisibleIndex.max)) {
			return;
		}

		const newVisiblePanels = this.calcVisiblePanels();
			
		this.updatePanelFromDOM(newVisiblePanels);
		this.updatePanelPosition(newVisiblePanels);
		
		this.visiblePanels = newVisiblePanels;
	}
}
```

And I've split function as follow

```ts
private getLastPanels() {
	const isHorizontal = this.options.horizontal;
	const panelManager = this.panelManager;
	const allPanelCount = panelManager.getRange().max + 1;
	const cloneCount = panelManager.getCloneCount();
	const cameraPos = this.getCameraPosition();
	const basePanel = this.nearestPanel;
	const viewportBbox = this.getBbox();
	const bbox = this.options.overflow
		? this.getWindowBBox()
		: viewportBbox;
	const viewportPos = isHorizontal
		? viewportBbox.x
		: viewportBbox.y;

	const checkLastPanel = (
		panel: Panel,
		getNextPanel: (panel: Panel) => Panel | null,
		isOutOfViewport: (panel: Panel) => boolean,
	): Panel => {
		let lastPanel = panel;
		while (true) {
			const nextPanel = getNextPanel(lastPanel);
			if (!nextPanel || isOutOfViewport(nextPanel)) {
				break;
			}
			lastPanel = nextPanel;
		}
		return lastPanel;
	};

	const lastPanelOfNextDir = checkLastPanel(basePanel, panel => {
		const nextPanel = panel.nextSibling;

		if (!nextPanel || nextPanel.getPosition() < panel.getPosition()) {
			return null;
		} else {
			return nextPanel;
		}
	}, panel => {
		const panelOffset = panel.getPosition() - cameraPos;
		const bboxNextPos = isHorizontal
			? bbox.x + bbox.width
			: bbox.y + bbox.height;
		return viewportPos + panelOffset >= bboxNextPos;
	});

	const lastPanelOfPrevDir = checkLastPanel(basePanel, panel => {
		const prevPanel = panel.prevSibling;

		if (!prevPanel || prevPanel.getPosition() > panel.getPosition()) {
			return null;
		} else {
			return prevPanel;
		}
	}, panel => {
		const panelOffset = panel.getPosition() - cameraPos;
		const bboxPrevPos = isHorizontal
			? bbox.x
			: bbox.y;
		return viewportPos + panelOffset + panel.getSize() <= bboxPrevPos;
	});

	return {lastPanelOfNextDir, lastPanelOfPrevDir};
}

private getNewVisibleIndex(lastPanelOfPrevDir, lastPanelOfNextDir) {
	const minPanelCloneIndex = lastPanelOfPrevDir.getCloneIndex();
	const maxPanelCloneOffset = allPanelCount * (lastPanelOfNextDir.getCloneIndex() + 1);
	const minPanelCloneOffset = minPanelCloneIndex > -1
		? allPanelCount * (cloneCount - minPanelCloneIndex)
		: 0;

	const newVisibleIndex = {
		// Relative index including clone, can be negative number
		min: lastPanelOfPrevDir.getIndex() - minPanelCloneOffset,
		// Relative index including clone
		max: lastPanelOfNextDir.getIndex() + maxPanelCloneOffset,
	};

	// Stopped on first cloned first panel
	if (lastPanelOfPrevDir.getIndex() === 0 && lastPanelOfPrevDir.getCloneIndex() === 0) {
		newVisibleIndex.min = allPanelCount;
	}

	return newVisibleIndex;
}

private getAddedRemovedPanels(newVisiblePanels) {
	const prevVisiblePanels = this.visiblePanels;
	const prevRefCount = prevVisiblePanels.map(() => 0);
	const newRefCount = newVisiblePanels.map(() => 0);

	prevVisiblePanels.forEach((prevPanel, prevIndex) => {
		newVisiblePanels.forEach((newPanel, newIndex) => {
			if (prevPanel === newPanel) {
				prevRefCount[prevIndex]++;
				newRefCount[newIndex]++;
			}
		});
	});

	const removedPanels = prevRefCount.reduce((removed: Panel[], count, index) => {
		return count === 0
			? [...removed, prevVisiblePanels[index]]
			: removed;
	}, []);
	const addedPanels = newRefCount.reduce((added: Panel[], count, index) => {
		return count === 0
			? [...added, newVisiblePanels[index]]
			: added;
	}, []);

	return {removedPanels, addedPanels};
}

private updatePanelFromDOM(newVisiblePanels) {
	const {removedPanels, addedPanels} = getAddedRemovedPanels(newVisiblePanels);
	const cameraElement = this.cameraElement;

	removedPanels.forEach(panel => {
		const panelElement = panel.getElement();
		panelElement.parentNode && cameraElement.removeChild(panelElement);
	});

	const fragment = document.createDocumentFragment();
	addedPanels.forEach(panel => {
		fragment.appendChild(panel.getElement());
	});

	cameraElement.appendChild(fragment);
}

private updatePanelPosition(newVisiblePanels) {
	if (newVisiblePanels.length > 0) {
		const firstPanelPos = panelManager.firstPanel()!.getPosition();
		const firstVisiblePanelPos = newVisiblePanels[0].getPosition();
		if (firstVisiblePanelPos >= firstPanelPos) {
			state.positionOffset = firstVisiblePanelPos - firstPanelPos;
		}
	}

	newVisiblePanels.forEach(panel => {
		panel.setPositionCSS(state.positionOffset);
	});
}
```

Reply:
Neat, but it might be not good for `Viewport` to increase methods count as it already has a bunch of them.
I think we should split our modules into separate files later.

Reply:
I agree `viewport` has too many method and it should be splited each other. But I think  if separating them in advance, we might find it convenient to refactor them later. And  I currently can get the advantage that I can read more comfortably .

Reply:
Sure, I'll split those functions.
It's okay to remember that we should refactor in one day.